include/larch/custom_file_io.hpp

```diff
@@ -0,0 +1,373 @@
+#pragma once
+
+#include <iostream>
+#include <fstream>
+#include <vector>
+#include <unordered_map>
+
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <boost/iostreams/filtering_stream.hpp>
+#include <boost/iostreams/filter/gzip.hpp>
+#include <boost/iostreams/copy.hpp>
+
+#include "larch/subtree/subtree_weight.hpp"
+#include "larch/subtree/parsimony_score.hpp"
+#include "larch/dag_loader.hpp"
+#include "proto/dag.pb.h"
+#include "parsimony.pb.h"
+#include "larch/newick.hpp"
+
+class CustomFileIO {
+ public:
+  struct Header {
+    size_t node_count = 0;
+    size_t edge_count = 0;
+    size_t leaf_count = 0;
+  };
+
+  static MADAGStorage<> ReadDAG(std::string_view path) {
+    std::ifstream infile(std::string{path}, std::ios::binary);
+
+    MADAGStorage<> dag_storage = MADAGStorage<>::EmptyDefault();
+    auto dag = dag_storage.View();
+    Header header;
+
+    while (!infile.eof()) {
+      auto append_id = ReadData<char>(infile);
+      auto nxt_offset = ReadData<std::streampos>(infile);
+      if (infile.eof()) {
+        break;
+      }
+
+      std::string test_str;
+      switch (append_id) {
+        case 'H':
+          header = ReadHeader(infile, dag);
+          break;
+        case 'R':
+          ReadReferenceSequence(infile, dag);
+          break;
+        case 'N':
+          ReadNodes(infile, dag);
+          break;
+        case 'E':
+          ReadEdges(infile, dag);
+          break;
+        case 'X':
+          test_str = ReadString(infile);
+          std::cout << ">>> READ test: '" << test_str << "'" << std::endl;
+          break;
+        default:
+          std::cerr << "ERROR: Invalid append_id '" << append_id << "'" << std::endl;
+          std::exit(EXIT_FAILURE);
+      }
+      Assert(nxt_offset == infile.tellg());
+    }
+    infile.close();
+
+    dag.BuildConnections();
+    dag.AssertUA();
+    for (auto node : dag.GetNodes()) {
+      if (node.IsLeaf() and not node.HaveSampleId()) {
+        node = SampleId{node.GetCompactGenome().ToString()};
```

Comment:
I think there needs to be a dag.RecomputeCompactGenomes() call before this. Nodes are written without cgs, and are recovered from the edge mutations.